File: src/main/python/org/o3project/odenos/remoteobject/remote_object_manager.py
# -*- coding:utf-8 -*-

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #

import logging
import copy
from urlparse import urlparse

from org.o3project.odenos.remoteobject.remote_object import RemoteObject
from org.o3project.odenos.remoteobject.object_property import ObjectProperty
from org.o3project.odenos.remoteobject.message.request import Request
from org.o3project.odenos.remoteobject.message.response import Response
from org.o3project.odenos.remoteobject.manager.component.event.component_changed import(
    ComponentChanged
)
from org.o3project.odenos.remoteobject.manager.system.event.component_manager_changed\
    import ComponentManagerChanged
from org.o3project.odenos.core.util.request_parser import RequestParser


class RemoteObjectManager(RemoteObject):

    # FIXME
    DESCRIPTION = "python's RemoteObjectManager"
    REMOTE_OBJECT_TYPES = "remote_object_types"

    def __init__(self, object_id, dispatcher):
        RemoteObject.__init__(self, object_id, dispatcher)
        self.remote_object_classes = {}
        self.remote_objects = {}
        self.dispatcher = dispatcher
        self._parser = RequestParser()
        self._add_rules()
        self._object_property.set_property(RemoteObjectManager.REMOTE_OBJECT_TYPES, "")

    def register_remote_objects(self, remote_objects):
        for remote_object in remote_objects:
            remote_object_name = remote_object.__name__
            logging.debug("Register RemoteObject Type:%s", remote_object_name)

            self.remote_object_classes[remote_object_name] = remote_object

        types = ",".join(self.remote_object_classes.keys())
        self._object_property.set_property(RemoteObjectManager.REMOTE_OBJECT_TYPES,
                                           types)

    def _add_rules(self):
        rules = []
        rules.append({RequestParser.PATTERN: r"^remote_object_types/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_remote_object_types,
                      RequestParser.PARAMS: 0})
        rules.append({RequestParser.PATTERN: r"^remote_objects/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_remote_objects,
                      RequestParser.PARAMS: 0})
        rules.append({RequestParser.PATTERN: r"^remote_objects/"
                      + "([a-zA-Z0-9_-]+)/?$",
                      RequestParser.METHOD: Request.Method.PUT,
                      RequestParser.FUNC: self._do_put_remote_object,
                      RequestParser.PARAMS: 2})
        rules.append({RequestParser.PATTERN: r"^remote_objects/"
                      + "([a-zA-Z0-9_-]+)/?$",
                      RequestParser.METHOD: Request.Method.GET,
                      RequestParser.FUNC: self._do_get_remote_object,
                      RequestParser.PARAMS: 1})
        rules.append({RequestParser.PATTERN: r"^remote_objects/"
                      + "([a-zA-Z0-9_-]+)/?$",
                      RequestParser.METHOD: Request.Method.DELETE,
                      RequestParser.FUNC: self._do_delete_remote_object,
                      RequestParser.PARAMS: 1})
        self._parser.add_rule(rules)

    def _on_request(self, request):
        return self._parser.action(request)

    def _do_get_remote_object_types(self):
        return Response(Response.StatusCode.OK, self.remote_object_classes.keys())

    def _do_get_remote_objects(self):
        body = {}
        for object_id in self.remote_objects:
            body[object_id] = \
                self.remote_objects[object_id].object_property.packed_object()
        return Response(Response.StatusCode.OK, body)

    def _do_get_remote_object(self, object_id):
        if object_id in self.remote_objects:
            return Response(Response.StatusCode.OK,
                            self.remote_objects[object_id].object_property)
        return Response(Response.StatusCode.NOT_FOUND, None)

    def _do_put_remote_object(self, obj_prop, object_id):
        remote_object_type = obj_prop[ObjectProperty.OBJECT_TYPE]

        if remote_object_type not in self.remote_object_classes:
            return Response(Response.StatusCode.BAD_REQUEST, None)
        elif object_id in self.remote_objects:
            return Response(Response.StatusCode.CONFLICT, None)

        remote_object_class = self.remote_object_classes[remote_object_type]
        self.remote_objects[object_id] = remote_object_class(object_id, self.dispatcher)

        if self.remote_objects[object_id].on_initialize(obj_prop):
            self.remote_objects[object_id].set_state(ObjectProperty.State.RUNNING)
        else:
            self.remote_objects[object_id].set_state(ObjectProperty.State.ERROR)

        logging.debug("Created RemoteObject Type:%s ID:%s", remote_object_type, object_id)
        return Response(Response.StatusCode.CREATED,
                        self.remote_objects[object_id].object_property)

    def _do_delete_remote_object(self, object_id):
        if object_id in self.remote_objects:
            remote_object = self.remote_objects[object_id]
            prev = copy.deepcopy(remote_object._object_property).packed_object()
            remote_object.on_finalize()
            del self.remote_objects[object_id]
            logging.debug("Deleted RemoteObject ID:%s", object_id)

        return Response(Response.StatusCode.OK, None)

File: src/main/java/org/o3project/odenos/remoteobject/RemoteObjectManager.java
/*
 * Copyright 2015 NEC Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.o3project.odenos.remoteobject;

import org.apache.commons.lang.StringUtils;
import org.apache.zookeeper.CreateMode;
import org.o3project.odenos.remoteobject.message.MessageBodyUnpacker.ParseBodyException;
import org.o3project.odenos.remoteobject.message.Request;
import org.o3project.odenos.remoteobject.message.Response;
import org.o3project.odenos.remoteobject.messagingclient.MessageDispatcher;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;
import org.o3project.odenos.core.logging.message.LogMessage;

import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

public class RemoteObjectManager extends RemoteObject {
  private static final Logger log = LogManager.getLogger(RemoteObjectManager.class);

  public static final String ATTR_OBJTYPE = "remote_object_types";
  public static final String ZK_CMPMGR_PATH = "/component_managers";
  public static final String ZK_CMP_PATH = "/components";
  private static final String OBJ_ID = "obj_id";

  protected RequestParser<IActionCallback> parser;
  protected Map<String, Class<? extends RemoteObject>> classList =
      new HashMap<String, Class<? extends RemoteObject>>();
  protected Map<String, RemoteObject> objects = new HashMap<String, RemoteObject>();

  /**
   * Constructor.
   * @param id Remote Object id.
   * @param disp Message Dispatcher id.
   */
  public RemoteObjectManager(String id, MessageDispatcher disp) {
    super(id, disp);
    this.objectProperty.setProperty(ATTR_OBJTYPE, "");
    parser = this.createParser();
  }

  /**
   * register Remote Objects.
   * @param classes Remote Object classes.
   */
  public void registerRemoteObjects(Set<Class<? extends RemoteObject>> classes) {
    for (Class<? extends RemoteObject> clazz : classes) {
      String type = clazz.getSimpleName();
      if (!classList.containsKey(type)) {
        classList.put(type, clazz);
      }
    }

    this.objectProperty.setProperty(ATTR_OBJTYPE, StringUtils.join(classList.keySet(), ","));
  }

  @Override
  protected Response onRequest(Request request) {
    LogMessage.setSavedTxid(request.txid);
    log.debug(LogMessage.buildLogMessage(LogMessage.getSavedTxid(), "onRequest: {}, {}", request.method, request.path));

    RequestParser<IActionCallback>.ParsedRequest parsed = parser.parse(request);
    Response response = null;

    IActionCallback callback = parsed.getResult();
    if (callback == null) {
      response = new Response(Response.BAD_REQUEST, null);
      LogMessage.delSavedTxid();
      return response;
    }
    try {
      response = callback.process(parsed);
    } catch (Exception e) {
      log.error(LogMessage.buildLogMessage(LogMessage.getSavedTxid(), "Exception Request: {}, {}", request.method, request.path));
      response = new Response(Response.BAD_REQUEST, null);
    }
    if (response == null) {
      response = new Response(Response.BAD_REQUEST, null);
    }
    LogMessage.delSavedTxid();
    return response;
  }

  protected RequestParser<IActionCallback> createParser() {
    return new RequestParser<IActionCallback>() {
      {
        addRule(Request.Method.GET,
            "remote_object_types",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return getRemoteObjectTypes();
              }
            });
        addRule(Request.Method.GET,
            "objects",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return getRemoteObjects();
              }
            });
        addRule(Request.Method.GET,
            "objects/<obj_id>",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return getRemoteObject(parsed.getParam(OBJ_ID));
              }
            });
        addRule(Request.Method.PUT,
            "objects/<obj_id>",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed)
                  throws ParseBodyException {
                return createRemoteObject(
                    parsed.getParam(OBJ_ID),
                    parsed.getRequest().getBody(ObjectProperty.class));
              }
            });
        addRule(Request.Method.DELETE,
            "objects/<obj_id>",
            new IActionCallback() {
              @Override
              public Response process(
                  final RequestParser<IActionCallback>.ParsedRequest parsed) {
                return deleteRemoteObject(parsed.getParam(OBJ_ID));
              }
            });
      }
    };
  }

  protected Response getRemoteObjectTypes() {
    return new Response(Response.OK, new ArrayList<String>(classList.keySet()));
  }

  protected Response getRemoteObjects() {
    Map<String, ObjectProperty> map = new HashMap<String, ObjectProperty>();
    for (Entry<String, RemoteObject> e : this.objects.entrySet()) {
      map.put(e.getKey(), e.getValue().getProperty());
    }
    return new Response(Response.OK, map);
  }

  protected Response getRemoteObject(String id) {
    RemoteObject object = this.objects.get(id);
    if (object == null) {
      return new Response(Response.NOT_FOUND, null);
    }
    return new Response(Response.OK, object.getProperty());
  }

  protected Response createRemoteObject(final String id, final ObjectProperty prop) {
    if (prop == null) {
      return new Response(Response.BAD_REQUEST, "Bad format: ObjectProperty is expected");
    }

    if (!classList.containsKey(prop.getObjectType())) {
      return new Response(Response.BAD_REQUEST, "Error unknown type ");
    }

    if (objects.containsKey(id)) {
      return new Response(Response.CONFLICT, "the RemoteObject is already created");
    }
    prop.setProperty(ObjectProperty.PropertyNames.OBJECT_ID, id);

    RemoteObject object = null;
    try {
      Class<? extends RemoteObject> clazz = classList.get(prop.getObjectType());
      try {
        Constructor<? extends RemoteObject> constructor =
            clazz.getConstructor(String.class, MessageDispatcher.class);
        object = constructor.newInstance(id, this.getMessageDispatcher());
      } catch (NoSuchMethodException e) {
        Constructor<? extends RemoteObject> constructor =
            clazz.getConstructor(String.class, String.class, MessageDispatcher.class);
        object = constructor.newInstance(
            id, objectProperty.getBaseUri(), this.getMessageDispatcher());
      }
      objects.put(id, object);
      if (object.onInitialize(prop)) {
        object.setState(ObjectProperty.State.RUNNING);
      } else {
        object.setState(ObjectProperty.State.ERROR);
      }
      zkCreatePath(ZK_CMP_PATH + "/" + getObjectId() + ":" +id,
          CreateMode.EPHEMERAL);
    } catch (Exception e) {
      return new Response(Response.INTERNAL_SERVER_ERROR, e.getMessage());
    }
    return new Response(Response.CREATED, object.getProperty());
  }

  protected Response deleteRemoteObject(final String id) {
    if (objects.containsKey(id)) {
      RemoteObject object = objects.get(id);
      ObjectProperty prev = (ObjectProperty) object.getProperty().clone();
      object.onFinalize();
      objects.remove(id);
      zkDeletePath(ZK_CMP_PATH + "/" + getObjectId() + ":" +id);
    }
    return new Response(Response.OK, null);
  }
}
